design/design-doc-composition-functions.md

```diff
@@ -0,0 +1,583 @@
+  things like ArgoCD and OPA.
+
+## Goals
+
+The proposal put forward by this document should:
+
+* Let folks use their composition tool and/or programming language of choice.
+* Support 'advanced' composition logic such as loops and conditionals.
+* Balance safety (e.g. sandboxing) with speed and simplicity.
+* Be possible to introduce behind a feature flag that is off by default.
+
+While not an explicit goal, it would also be ideal if the solution put forth by
+this document could serve as a test bed for new features in the contemporary
+'resources array' based form of Composition.
+
+## Proposal
+
+### Overview
+
+This document proposes that a new `functions` array be added to the existing
+`Composition` type. This array of functions would be called either instead of or
+in addition to the existing `resources` array in order to determine how an XR
+should be composed.
+
+```yaml
+apiVersion: apiextensions.crossplane.io/v2alpha1
+kind: Composition
+metadata:
+  name: example
+spec:
+  compositeTypeRef:
+    apiVersion: database.example.org/v1alpha1
+    kind: XPostgreSQLInstance
+  functions:
+  - name: my-cool-function
+    type: Container
+    container:
+      image: xkpg.io/my-cool-function:0.1.0
+```
+
+Under this proposal each function is the entrypoint of an OCI image, though the
+API is designed to support different function implementations (such as webhooks)
+in th future. The updated API would affect only the `Composition` type - no
+changes would be required to the schema of `CompositeResourceDefinitions`, XRs,
+etc.
+
+Notably the functions would not be responsible for interacting with the API
+server to create, update, or delete composed resources. Instead they instruct
+Crossplane which resources should be created, updated, or deleted.
+
+Under the proposed design functions could also be used for purposes besides
+rendering composed resources, for example validating the results of the
+`resources` array or earlier functions in the `functions` array. Furthermore, a
+function could also be used to implement 'side effects' such as triggering a
+replication or backup.
+
+Below is a more detailed example of an entry in the `functions` array.
+
+```yaml
+apiVersion: apiextensions.crossplane.io/v2alpha1
+kind: Composition
+metadata:
+  name: example
+spec:
+  compositeTypeRef:
+    apiVersion: database.example.org/v1alpha1
+    kind: XPostgreSQLInstance
+  functions:
+  - name: my-cool-function
+    type: Container
+    # Configuration specific to `type: Container` functions.
+    container:
+      # The OCI image to pull and run.
+      image: xkpg.io/my-cool-function:0.1.0
+    # An x-kubernetes-embedded-resource RawExtension (i.e. an unschemafied)
+    # Kubernetes resource. Passed to the function as the functionConfig block of
+    # its ResourceList.
+    config:
+      apiVersion: database.example.org/v1alpha1
+      kind: Config
+      metadata:
+        name: cloudsql
+      spec:
+        version: POSTGRES_9_6
```

Comment:
@turkenh @ron1 this is how I imagine per-function configuration would work. The `config` block specified here would be passed as the `functionConfig` to the relevant function. Therefore the input to each function is:

1. The `items` array - i.e. the XR and any composed resources.
2. The `functionConfig`, as specified in the `Composition`.

Taking the Helm chart based "reference function" @sergenyalcin is working on in https://github.com/crossplane/crossplane/issues/3067 for example, I imagine:

1. Which Helm chart to use would be specified in the `functionConfig`.
2. The inputs to the Helm chart (i.e. the `values.yaml`) would be the `ResourceList`, or at least the XR from said list.

Note that this assumes we're using a Helm chart expressly written to do the job of deriving composed resources from an XR; not a generic Helm chart. We already have `provider-helm` for rendering generic Helm charts.

One gap in this design is passing sensitive details to functions. As @turkenh pointed out a function might require credentials to operate, for example a username and password for a Helm repository. In this design those sensitive credentials would have to be provided inline in the Composition.

We could fairly easily allow config to be loaded from a Kubernetes Secret, for example:

```yaml
apiVersion: apiextensions.crossplane.io/v2alpha1
 kind: Composition
 metadata:
   name: example
 spec:
   compositeTypeRef:
     apiVersion: database.example.org/v1alpha1
     kind: XPostgreSQLInstance
   functions:
   - name: my-cool-function
     type: Container
     container:
       # The OCI image to pull and run.
       image: xkpg.io/my-cool-function:0.1.0
     configSource: Secret
     # Mutually exclusive with the config block.
     configSecretRef:
       namespace: crossplane-system
       name: my-cool-function-config
```

In this example the entire referenced `Secret` would be passed as the `functionConfig`.

I'm not sure just supporting Kubernetes secrets is really enough though. We've already begun adding support to Crossplane for other secret stores such as Vault, so I imagine it's just as likely someone would want to load function config from them. @turkenh suggested we could pass in connection details, but in that case:

1. Where are the connection details coming from? The XR? What sets them if rendering the XR is dependent on the function?
2. What if we need to pass in connection details _and_ other function config?

I'm tempted to leave this out of scope for the first pass of this feature and address it later once we have a few more concrete use cases.

Reply:
In the above example, the `my-cool-function` accepts parameter `spec.version`. Rather than hardcoding this parameter into the functionConfig, I would like to capture this value in the XR and then have it patched into the functionConfig. Is it possible to set the functionConfig parameter values in this dynamic fashion?

Reply:
> Rather than hardcoding this parameter into the functionConfig, is it possible to capture this value in the XR and then have it patched into the functionConfig? WDYT about setting functionConfig parameter values in this dynamic fashion?

Why pass it in via the `functionConfig` when the XR itself will always be available in the `items` array?

Reply:
Most kpt functions are reusable. An XRM function that depends on a specific XR schema will not be reusable. So providing an easy way to map XR field values to functionConfig values seems important to enable reusable functions.

The functions in the kpt catalog typically accept a few simple inputs. I would expect a catalog of XRM functions to be similar. Many XRs have custom types which prevents them from being directly consumed from reusable XRM functions.

Reply:
> Most kpt functions are reusable

I am assuming most XRM functions - say ~70% - will not be reusable. The primary motivation behind this design is to provide a way to take an XR as input and output composed resources. I'd consider "generic" functions (e.g. those for validation etc) to be a secondary concern.

Perhaps there are use cases I am not seeing here - I'd be happy to hear them.

Reply:
For example, say an existing reasonably complex XR has a spec field and value that I want to add as a label to all my composed resources. Kpt function https://catalog.kpt.dev/set-labels/v0.1/ uses the functionConfig to accomplish this. A reusable XRM function could do likewise. I would not want to rewrite this set-labels function for every different source XR. Rather, I would like to pick the field value out of the XR and patch it into the set-labels functionConfig to allow my XR and function to remain decoupled.

Reply:
I'm not opposed to a feature that patches from XR -> function config to allow for functions that are unaware of the XR's schema. That said, my inclination is to avoid building that in the first pass. We can add it later once clear use cases arise.